lex.go

```diff
@@ -0,0 +1,653 @@
+package mf2
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"unicode/utf8"
+)
+
+// eof is the end of file item.
+const eof = -1
+
+// itemType is the type of an item.
+type itemType int
+
+const (
+	itemError itemType = iota
+	itemEOF
+	itemVariable
+	itemFunction
+	itemExpressionOpen
+	itemExpressionClose
+	itemQuotedPatternOpen
+	itemQuotedPatternClose
+	itemText
+	itemKeyword
+	itemLiteral
+	itemOption
+	itemWhitespace
+	itemReserved
+	itemOperator
+	itemPrivate
+)
+
+// String returns a string representation of the item type.
+func (t itemType) String() string {
+	switch t {
+	default:
+		return "unknown"
+	case itemError:
+		return "error"
+	case itemEOF:
+		return "eof"
+	case itemVariable:
+		return "variable"
+	case itemFunction:
+		return "function"
+	case itemExpressionOpen:
+		return "expression open"
+	case itemExpressionClose:
+		return "expression close"
+	case itemQuotedPatternOpen:
+		return "quoted pattern open"
+	case itemQuotedPatternClose:
+		return "quoted pattern close"
+	case itemText:
+		return "text"
+	case itemKeyword:
+		return "keyword"
+	case itemLiteral:
+		return "literal"
+	case itemOption:
+		return "option"
+	case itemWhitespace:
+		return "whitespace"
+	case itemReserved:
+		return "reserved"
+	case itemOperator:
+		return "operator"
+	case itemPrivate:
+		return "private"
+	}
+}
+
+// Keywords.
+const (
+	keywordMatch = "match"
+	keywordLocal = "local"
+	keywordInput = "input"
+)
+
+// item is a token returned by the lexer.
+type item struct {
+	val string
+	typ itemType
+}
+
+// mk creates a new item with the given type and value.
+func mk(typ itemType, val string) item {
+	return item{typ: typ, val: val}
+}
+
+// lex creates a new lexer for the given input string.
+func lex(input string) *lexer { return &lexer{input: input} }
+
+// lexer is a lexical analyzer for MessageFormat2.
+//
+// See https://github.com/unicode-org/message-format-wg/blob/main/spec/syntax.md
```

Comment:
Since we can't follow any versions, perhaps we could add information about the date when syntax was used to build the implementation.

Reply:
This is something I'm trying to figure out. It's not gonna be the date as it is not deterministic.

Either git commit value or copy of an ABNF file in this repo.

Reply:
just link to abnf from commit this was made from, should be good
e.g.

https://github.com/unicode-org/message-format-wg/blob/7c00820a0462679eba696181c45bfadb43d2eedd/spec/message.abnf